### PR TITLE
Create UUID hierarchy

### PR DESCRIPTION
**Pull request by alejandrohdezma**
# What has been done in this PR?

Convert `UUID` class into a `sealed trait` + class hierarchy.

The following sub-classes have been created:

- `UUID.Nil`: For the `UUID(0, 0)` ([null UUID](https://tools.ietf.org/html/rfc4122#section-4.1.7))
- `UUID.V1`: [For the version-1 UUID](https://tools.ietf.org/html/rfc4122#section-4.1.3)
- `UUID.V2`: [For the version-2 UUID](https://tools.ietf.org/html/rfc4122#section-4.1.3)
- `UUID.V3`: [For the version-3 UUID](https://tools.ietf.org/html/rfc4122#section-4.1.3)
- `UUID.V4`: [For the version-4 UUID](https://tools.ietf.org/html/rfc4122#section-4.1.3)
- `UUID.V5`: [For the version-5 UUID](https://tools.ietf.org/html/rfc4122#section-4.1.3)
- `UUID.UnknownVersion`: For those UUIDs that don't fall in any of the other classes

Also the following convenient methods have been created:

- `as[A <: UUID]: Option[A]`: allows narrowing a `UUID` into the specified version

```scala
// Create V4 UUID
val uuid = UUID.fromString("7d3e5c3e-c3c1-4c3c-957d-15a454f8d376").asScala

uuid.as[UUID.V4] // Some(UUID("7d3e5c3e-c3c1-4c3c-957d-15a454f8d376"))
uuid.as[UUID.V1] // None
```

- `is[A <: UUID]: Boolean`: checks if the `UUID` version is the provided one

```scala
// Create V4 UUID
val uuid = UUID.fromString("7d3e5c3e-c3c1-4c3c-957d-15a454f8d376").asScala

uuid.is[UUID.V4] // true
uuid.is[UUID.V1] // false
```

### CODE REVIEW

**Comment by purrgrammer** [comment]
Amazing job :clap:

File: src/main/scala/memeid/UUID.scala [hunk 1]
```diff
@@ -46,29 +82,79 @@ final class UUID private[memeid] (private[memeid] val juuid: JUUID) {
...
+   */
+  case object Nil extends UUID { override val juuid: JUUID = new JUUID(0, 0) }
+
+  /**
+   * Version 1 UUIDs are those generated using a timestamp and the MAC address of the
+   * computer on which it was generated.
+   *
+   * @see [[https://tools.ietf.org/html/rfc4122#section-4.1.3]]
+   */
+  final class V1 private[memeid] (override private[memeid] val juuid: JUUID) extends UUID
```
Comment (gutiory) [inline src/main/scala/memeid/UUID.scala]: What's the `class V1 private[memeid]` effect? . Does it make `V1` private outside `memeid` package?

Comment (alejandrohdezma) [reply src/main/scala/memeid/UUID.scala]: Nope, it makes `V1` constructor private outside `memeid`, so no one (but us) can create `V1` UUIDs (to force users to use our convenient methods)

Comment (gutiory) [reply src/main/scala/memeid/UUID.scala]: Amazing!. Thanks!

**Pull request merged**